gossip/src/cluster_info.rs

```diff
@@ -670,7 +691,11 @@ impl ClusterInfo {
     }
 
     pub fn id(&self) -> Pubkey {
-        self.keypair.read().unwrap().pubkey()
+        // `self.keypair.read().unwrap().pubkey()` is more straight-forward to use here.
+        // However, self.keypair could be dummy in some very odd situation
+        // (i.e. ledger-tool's simulate-leader-production). So, use `self.my_contact_info` here.
+        // Other than the edge case, both are equivalent.
+        *self.my_contact_info.read().unwrap().pubkey()
```

Comment:
@behzadnouri please let me know if you disagree with the changes as justified by the source code comment.

Reply:
I pretty much would rather avoid introducing scenarios where`contact_info.keypair.pubkey() != contact_info.pubkey`.

Can you please provide more context why we need a `ClusterInfo` with a `contact_info` which we do not own the `keypair`? To me this seems pretty error-prone and I would much prefer we try an alternative.

Reply:
Agree with @behzadnouri here. It seems the reason we are adding this new inconsistency is because `BankingStage` takes it as an argument.

It'd be better for us to refactor `BankingStage` to not use `ClusterInfo` imo.

We use `ClusterInfo` for 2 things:

1. Creating the `Forwarder`
    - could pass an `Option<Forwarder>` as arg to `BankingStage` instead
    - ^ would need some changes to rip out mandatory forwarding in tlmi / voting threads
    - alternatively could make forwarder an enum w/ disabled variant or even a trait instead
2. Getting validator id for checking if we are leader
    - easily can pass `Pubkey` instead

Reply:
hmm, my dcou based hack is unpopular.. ;) I did my part of little hassle. how about this?: 2b33131bec622c09d4254ee25727b3de764709fd

> > Can you please provide more context why we need a `ClusterInfo` with a `contact_info` which we do not own the `keypair`?

> It seems the reason we are adding this new inconsistency is because `BankingStage` takes it as an argument.

@apfitzge 's understanding is correct. note that such broken ClusterInfo is only ever created under dcou code-path, though.

> Getting validator id for checking if we are leader
> 
>     * easily can pass `Pubkey` instead

Sadly, this isn't easy because the identity `Pubkey` can be hot-swapped inside `ClusterInfo`. That lead me to take this trait direction...:

> 1. alternatively could make forwarder ... a trait instead

Also, note that `BroadcastStageType` also needs `ClusterInfo`. Fortunately, it seems the new hacky trait `LikeClusterInfo` plumbing isn't needed for it.

That said, I wonder this additional production code is worth to maintain, only to support some obscure development `ledger-tool` subcommand. Anyway, I'm not that opinionated. I just want to merge this pr.